### DISCUSSION TITLE
How does collaboration work?

### DISCUSSION THREAD

**Discussion by Immortalin** [uxbox/uxbox#679] [Q&A]
I looked through the code base and I could not find any CRDT data structures nor could I find operational transform code. How does the collaboration aspect work? The only relevant file that I can find is [this](https://github.com/penpot/penpot/blob/ef9b0062dd1377e825360c7a7d35c763674832cf/frontend/src/app/main/ui/workspace/sidebar/history.cljs).

**Comment by niwinz** [comment]
Hi @Immortalin 
The collaborative edition is implemented using something like event-sourcing-like approach that is already used for perform all modifications to a file, undo history and many other things. The `file` is a central point of contention where all changes are serialized. The relevant file is https://github.com/penpot/penpot/blob/develop/common/app/common/pages/changes.cljc

**Reply by Immortalin** [reply]
The changes will be serialised correctly even in the case of conflicting changes? They will be eventually consistent?

I understand Figma's approach looks like this:
https://www.figma.com/blog/how-figmas-multiplayer-technology-works/

**Reply by niwinz** [reply]
The conflict is managed at shape attrbute level (much in the same line as figma); if two users modify the same attrubute, the lasts one wins; If two users modify different attributes, they both should success. On the backend side, there are only one operation at  time updating file, any other operation are queued and processed  one by one, ensuring consistency on "user action transaction". The user action transaction can consist on a single change or a  collection of changes. 

I think that in a big picture the current approach is very similar to the implemented in figma. But right now, i a very young impl and probably has some issues; but for now it works for us.

**Comment by hirunatan** [comment]
We have also just published a documentation site. In the section about architecture, we talk about the websocket and broker structure, that is used to communicate changes between users editing the same file.

https://help.penpot.app/developer-guide/architecture/#broker
